spec/java/util/StringBuilder.main.lsl

```diff
@@ -0,0 +1,1022 @@
+        result = action CALL_METHOD(this.storage, "charAt", [index]);
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.setLength (@target self: StringBuilder, newLength: int): void
+    {
+        if (newLength < 0)
+        {
+            //val message: String = "String index out of range: " + action OBJECT_TO_STRING(newLength);
+            action THROW_NEW("java.lang.StringIndexOutOfBoundsException", []);
+        }
+        var i: int = 0;
+        val newStr: array<char> = action ARRAY_NEW("char", newLength);
+        action LOOP_FOR(i, 0, newLength, +1, _setNewLength_loop(i, newStr));
+
+        this.storage = action OBJECT_TO_STRING(newStr);
+        this.length = newLength;
+    }
+
+
+    @Phantom proc _setNewLength_loop (i: int, newStr: array<char>): void
+    {
+        if (i < this.length)
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [i]);
+        // TODO: problem with: ''
+        //else
+            // That's right ? Did I understand correctly ?
+        //    newStr[i] = '0';
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.setCharAt (@target self: StringBuilder, index: int, ch: char): void
+    {
+        _checkIndex(index);
+        val newStr: array<char> = action ARRAY_NEW("char", this.length);
+        var i: int = 0;
+        action LOOP_FOR(i, 0, this.length, +1, _setNewLength_loop(i, index, newStr, ch));
+        this.storage = action OBJECT_TO_STRING(newStr);
+    }
+
+
+    @Phantom proc _setNewLength_loop (i: int, index: int, newStr: array<char>, ch: char): void
+    {
+        if(i != index)
+            newStr[i] = action CALL_METHOD(this.storage, "charAt", [i]);
+        else
+            newStr[i] = ch;
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.trimToSize (@target self: StringBuilder): void
+    {
+
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.substring (@target self: StringBuilder, start: int): String
+    {
+        result = _substring(start, this.length);
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.substring (@target self: StringBuilder, start: int, end: int): String
+    {
+        result = _substring(start, end);
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.subSequence (@target self: StringBuilder, start: int, end: int): CharSequence
+    {
+        _checkRangeSIOOBE(start, this.length, this.length);
+        val sizeNewString: int = end - start;
+        val newStr: array<char> = action ARRAY_NEW("char", sizeNewString);
+
+        var i: int = 0;
+        action LOOP_FOR(i, start, end, +1, _newSubString_loop(i, newStr));
+        result = action OBJECT_TO_STRING(newStr);
+    }
+
+
+    // within java.lang.AbstractStringBuilder
+    fun *.getChars (@target self: StringBuilder, srcBegin: int, srcEnd: int, dst: array<char>, dstBegin: int): void
+    {
+        _checkRangeSIOOBE(srcBegin, srcEnd, this.length);
+        var n: int = srcEnd - srcBegin;
```

Comment:
Can we have a suggestion for the USVM here? What important properties we know about the value of this variable?